### PR TITLE
Added cmake build steps to support spack package

### PR DESCRIPTION
**Pull request by dyokelson**
Pull request for adding a cmake build for AMG2023. Also incorporates Caliper instrumentation and Adiak metadata (thus supersedes Tre's PR #2). Will notify when code is ready for review. @slabasan, @daboehme, @pearce8

### CODE REVIEW

File: amg.c [hunk 1]
```diff
@@ -362,7 +390,6 @@ main( hypre_int argc,
    /*-----------------------------------------------------------
     * Set up the RHS and initial guess
     *-----------------------------------------------------------*/
-
```
Comment (slabasan) [inline amg.c:365]: restore line

File: amg.c [hunk 2]
```diff
@@ -421,18 +448,23 @@ main( hypre_int argc,
       HYPRE_IJMatrixPrint(ij_A, "IJ.out.A");
       HYPRE_IJVectorPrint(ij_b, "IJ.out.b");
       HYPRE_IJVectorPrint(ij_x, "IJ.out.x0");
-
```
Comment (slabasan) [inline amg.c:424]: restore line

File: amg.c [hunk 3]
```diff
@@ -421,18 +448,23 @@ main( hypre_int argc,
...
    }
 
    /*-----------------------------------------------------------
     * Problem 2: Solve a 7pt 3D Laplace problem with AMG-PCG
     *-----------------------------------------------------------*/
 
-
```
Comment (slabasan) [inline amg.c:431]: restore line

File: variorum_power.json [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+          "select" :
+          [
+            { "expr": "max(max#power.max)", "as": "Power/node (max)", "unit": "Watts" },
+            { "expr": "avg(avg#power.avg)", "as": "Power/node (avg)", "unit": "Watts" }
+          ]
+        }
+    ]
+}
+]
+}
```
Comment (slabasan) [inline variorum_power.json:36]: We can remove this JSON file from the PR as we are still waiting on the variorum spack package.

**Review by dyokelson — COMMENTED** [review]
updates look good

File: amg.c [hunk 5]
```diff
@@ -337,7 +366,6 @@ main( hypre_int argc,
    /*-----------------------------------------------------------
     * Set up matrix
     *-----------------------------------------------------------*/
-
```
[the remark this answers is not in the document]
Comment (dyokelson) [reply amg.c:340]: done

File: amg.c (hunk 1 above)
Comment (dyokelson) [reply amg.c:365]: done

File: amg.c (hunk 3 above)
Comment (dyokelson) [reply amg.c:431]: done

File: amg.c [hunk 6]
```diff
@@ -160,12 +170,31 @@ main( hypre_int argc,
    /*-----------------------------------------------------------
     * Initialize MPI
     *-----------------------------------------------------------*/
-
+
```
[the remark this answers is not in the document]
Comment (dyokelson) [reply amg.c:173]: done

File: amg.c (hunk 2 above)
Comment (dyokelson) [reply amg.c:424]: done

File: variorum_power.json (hunk 4 above)
Comment (dyokelson) [reply variorum_power.json]: done

File: amg.c [hunk 7]
```diff
@@ -624,6 +709,9 @@ main( hypre_int argc,
       HYPRE_ParCSRGMRESDestroy(pcg_solver);
       FOM2 = cum_nnz_AP * (HYPRE_Real)num_iterations / wall_time;
```
Comment (slabasan) [inline amg.c:710]: Should this be included in the calculate-FOM region?

File: amg.c [hunk 8]
```diff
@@ -662,6 +758,10 @@ main( hypre_int argc,
    /* Finalize MPI */
    hypre_MPI_Finalize();
 
+#ifdef USE_CALIPER
+   CALI_MARK_END("main");
+   adiak_fini();
+#endif
```
Comment (daboehme) [inline amg.c:764]: The end main region and adiak_fini calls need to move before `hypre_MPI_Finalize()`

Comment (dyokelson) [reply amg.c:761]: done

File: amg.c (hunk 7 above)
Comment (dyokelson) [reply amg.c:710]: good question, it does seem like it should be. Something similar is happening on line 671 though. I wonder if it is just calculating the time, as opposed to an actual value. Sorry I didn't add these annotation locations so I'm not quite sure, maybe @daboehme can confirm?

**Comment by daboehme** [comment]
Looks good to me! The CMake version works for HIP, CUDA, and CPU only builds.

File: CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+if("${CMAKE_SOURCE_DIR}" STREQUAL "${CMAKE_BINARY_DIR}")
+   message(FATAL_ERROR "In-source builds are not supported. Please remove "
+                       "CMakeCache.txt from the 'src' dir and configure an "
+                       "out-of-source build in another directory.")
+endif()
+
+add_definitions(-DAMG_CMAKE_BUILD=1)
+set(CMAKE_INSTALL_RPATH_USE_LINK_PATH TRUE)
+
+option(AMG_WITH_MPI "Use MPI" FALSE)
```
Comment (daboehme) [inline CMakeLists.txt:21]: We probably want MPI on by default.

**Review by daboehme — APPROVED** [review]
Looks good!

File: amg.c [hunk 10]
```diff
@@ -598,19 +670,32 @@ main( hypre_int argc,
 
       FOM2 = cum_nnz_AP / wall_time;
```
Comment (ulrikeyang) [inline amg.c:671]: This was a bug and needs to be FOM1

File: amg.c [hunk 11]
```diff
@@ -598,19 +670,32 @@ main( hypre_int argc,
...
 
+#ifdef USE_CALIPER
+      CALI_MARK_BEGIN("setup-FOM");
+#endif
       if (myid == 0)
       {
          hypre_printf ("\nFOM_Setup: nnz_AP / Setup Phase Time: %e\n\n", FOM2);
```
Comment (ulrikeyang) [inline amg.c:678]: Same here: FOM1

File: amg.c [hunk 12]
```diff
@@ -598,19 +670,32 @@ main( hypre_int argc,
...
       }
 
+#ifdef USE_CALIPER
+      CALI_MARK_END("setup-FOM");
+      adiak_namevalue("Setup-FOM", adiak_general, NULL, "%f", FOM2);
```
Comment (ulrikeyang) [inline amg.c:683]: Then here FOM2 will also need to be replaced by FOM1.

File: CMakeLists.txt (hunk 9 above)
Comment (ulrikeyang) [reply CMakeLists.txt:21]: I think so, too.

Comment (dyokelson) [reply CMakeLists.txt]: Done

File: amg.c (hunk 11 above)
Comment (dyokelson) [reply amg.c]: Done

File: amg.c (hunk 12 above)
Comment (dyokelson) [reply amg.c]: Done

File: amg.c (hunk 10 above)
Comment (dyokelson) [reply amg.c]: Done

**Review by ulrikeyang — APPROVED** [review]
I think this is fine.

**Pull request merged**